File: Polars.FSharp/IO/ADO.NET/Read.fs
namespace Polars.FSharp

[<AutoOpen>]
module DataReaderRead =
    open System
    open Apache.Arrow
    open System.Collections.Generic
    open Polars.NET.Core.Arrow
    open Polars.NET.Core.Data
    open System.Data
    type LazyFrame with
        /// <summary>
        /// Lazily scan a sequence of objects using Apache Arrow Stream Interface.
        /// This supports predicate pushdown and streaming execution.
        /// Data is pulled from the sequence only when needed.
        /// </summary>
        /// <param name="data">The data source sequence.</param>
        /// <param name="batchSize">Rows per Arrow batch (default: 100,000).</param>
        /// <param name="useBuffered">Choose whether disk buffer file needed (for big data) <param>
        static member scanSeq<'T>(data: seq<'T>, ?batchSize: int, ?useBuffered: bool) : LazyFrame =
                let size = defaultArg batchSize 100_000
                let buffered = defaultArg useBuffered false

                // =========================================================
                // 1. Buffered Mode (Disk IPC)
                // =========================================================
                if buffered then
                    let scope = new IpcStreamService.TempIpcScope<'T>(data, size)
                    
                    // Get FileHandle
                    let handle = LazyFrame.ScanIpc(scope.FilePath).Handle
                    
                    { new LazyFrame(handle) with
                        member this.Dispose() =
                            base.Dispose()
                            scope.Dispose()
                    }

                // =========================================================
                // 2. Streaming Mode (Memory Safety & Lazy Fallback)
                // =========================================================
                else
                    let schema = ArrowConverter.GetSchemaFromType<'T>()

                    let streamFactory = Func<IEnumerable<RecordBatch>>(fun () ->
                        seq {
                            let mutable hasYielded = false
                            
                            let batches = ArrowConverter.ToArrowBatches(data, size).Prefetch()

                            for batch in batches do
                                hasYielded <- true
                                yield batch
                                batch.Dispose()
                            
                            if not hasYielded then
                                let emptyBatch = ArrowConverter.GetEmptyBatch<'T>()
                                yield emptyBatch
                                emptyBatch.Dispose()
                        }
                    )

                    let handle = ArrowStreamInterop.ScanStream(streamFactory, schema)
                    new LazyFrame(handle)

        /// <summary>
        /// Scan a database query lazily.
        /// Requires a factory function to create new IDataReaders for potential multi-pass scans.
        /// </summary>
        static member scanDb(readerFactory: unit -> IDataReader, ?batchSize: int, ?useBuffered: bool) : LazyFrame =
            let size = defaultArg batchSize 50_000
            let buffered = defaultArg useBuffered false

            // =========================================================
            // 1. Buffered Mode (Disk IPC)
            // =========================================================
            if buffered then
                let runBuffer () =
                    use reader = readerFactory()
                    new IpcStreamService.TempIpcScopeReader(reader, size)

                let scope = runBuffer()
                let handle = LazyFrame.ScanIpc(scope.FilePath).Handle

                { new LazyFrame(handle) with
                    member this.Dispose() =
                        base.Dispose()
                        scope.Dispose()
                }

            // =========================================================
            // 2. Streaming Mode (Memory)
            // =========================================================
            else
                // Probe Schema
                let schema = 
                    use reader = readerFactory()
                    ArrowTypeResolver.GetSchemaFromDataReader reader

                // Stream Factory
                let factory = Func<IEnumerable<RecordBatch>>(fun () ->
                    seq {
                        use reader = readerFactory()
                        let batches = DbToArrowStream.ToArrowBatches(reader, size).Prefetch()
                        
                        for batch in batches do
                            yield batch
                            batch.Dispose()
                    }
                )

                let handle = ArrowStreamInterop.ScanStream(factory, schema)
                new LazyFrame(handle)
        /// <summary>
        /// [Lazy][Streaming] Scan a database DataReader directly.
        /// <para>Upgraded to pure memory Streaming Mode using Arrow C-Data FFI!</para>
        /// </summary>
        static member scanDb(reader: IDataReader, ?batchSize: int) : LazyFrame =
            let size = defaultArg batchSize 50_000
            let schema = reader.GetArrowSchema()
            
            let stream = reader.ToArrowBatches(size).Prefetch() 
            
            let factory = Func<IEnumerable<RecordBatch>>(fun () -> stream)
            let handle = ArrowStreamInterop.ScanStream(factory, schema)
            
            new LazyFrame(handle)
    type DataFrame with
        /// <summary>
        /// [Eager] Create a DataFrame from an IDataReader (e.g. SqlDataReader).
        /// <para>
        /// Uses high-performance streaming ingestion via Apache Arrow.
        /// </para>
        /// </summary>
        /// <param name="reader">The open IDataReader instance.</param>
        /// <param name="batchSize">Number of rows per Arrow batch (default 50,000).</param>
        static member ReadDb(reader: IDataReader, ?batchSize: int) : DataFrame =
            let schema = reader.GetArrowSchema()

            let size = defaultArg batchSize 50_000
            
            let batchStream = reader.ToArrowBatches(size).Prefetch()
            
            let handle = ArrowStreamInterop.ImportEager(batchStream,schema)
            
            if handle.IsInvalid then

                let emptyBatch = new RecordBatch(schema, System.Array.Empty<IArrowArray>(), 0)

                let safeHandle = ArrowFfiBridge.ImportDataFrame emptyBatch
                new DataFrame(safeHandle)
            else
                new DataFrame(handle)

File: Polars.CSharp/IO/Database/ADO.NET/Read.cs
using System.Data;
using Apache.Arrow;
using Polars.NET.Core;
using Polars.NET.Core.Arrow;
using Polars.NET.Core.Data;

namespace Polars.CSharp;
public partial class DataFrame : IDisposable,IEnumerable<Series>,IPolarsDataFrame
{
    /// <summary>
    /// Create a DataFrame directly from a <see cref="IDataReader"/>.
    /// <para>
    /// This method streams data from the reader into Arrow batches, allowing for memory-efficient 
    /// loading of large datasets from databases (e.g., SQL Server, PostgreSQL, SQLite).
    /// </para>
    /// <para>
    /// It automatically maps C# types to Polars types (e.g., <see cref="decimal"/> to Decimal128, <see cref="DateTime"/> to Timestamp).
    /// </para>
    /// </summary>
    /// <param name="reader">The open IDataReader instance.</param>
    /// <param name="batchSize">The number of rows to process per Arrow batch. Default is 50,000.</param>
    /// <returns>A new DataFrame.</returns>
    /// <example>
    /// <code>
    /// // Mocking a DataTable as a data source
    /// var table = new System.Data.DataTable();
    /// table.Columns.Add("Product", typeof(string));
    /// table.Columns.Add("Price", typeof(decimal)); // Correctly maps to Polars Decimal128
    /// 
    /// table.Rows.Add("Laptop", 1234.56m);
    /// table.Rows.Add("Mouse", 99.99m);
    /// 
    /// using IDataReader reader = table.CreateDataReader();
    /// 
    /// var df = DataFrame.ReadDatabase(reader);
    /// df.Show();
    /// /* Output:
    /// shape: (2, 2)
    /// ┌─────────┬─────────────────────────┐
    /// │ Product ┆ Price                   │
    /// │ ---     ┆ ---                     │
    /// │ str     ┆ decimal[38,18]          │
    /// ╞═════════╪═════════════════════════╡
    /// │ Laptop  ┆ 1234.560000000000000000 │
    /// │ Mouse   ┆ 99.990000000000000000   │
    /// └─────────┴─────────────────────────┘
    /// */
    /// </code>
    /// </example>
    public static DataFrame ReadDatabase(IDataReader reader, int batchSize = 50_000)
    {
        // Get Schema 
        var schema = reader.GetArrowSchema();

        var batchEnumerable = reader.ToArrowBatches(batchSize).Prefetch();

        var handle = ArrowStreamInterop.ImportEager(batchEnumerable, schema);
        
        if (handle.IsInvalid)
        {
            var emptyBatch = new RecordBatch(schema, [], 0);
            return new DataFrame(ArrowFfiBridge.ImportDataFrame(emptyBatch));
        }
        
        return new DataFrame(handle);
    }
}
